compiler-core/src/type_/expression.rs

```diff
@@ -693,7 +693,17 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
             },
         );
 
-        Statement::Expression(call)
+        Statement::Use(Use {
+            call: Box::new(call),
+            location: use_.location,
+            right_hand_side_location: use_.right_hand_side_location,
+            assignments_location: use_.assignments_location,
+            // TODO: figure out what to do with assignments in the typed use
+            // case. They're not really needed so I'm discarding those but it
+            // doesn't feel great and sounds like it would be confusing if
+            // someone in the future tries to access those from the typed AST.
```

Comment:
I don't know what the two "it"s are in this sentence 😁

Reply:
Oops sorry 😆

So right now when we have to type a `use` expression we build an untyped function call and then let the typer type it. What we then do is we carry around the typed function call and use that for things like code generation, which makes it super simple.

The problem with this approach is that since what we're typing is a function call it's harder to get those typed patterns since they're just assignments we add to the body of the anonymous function passed by use.
There's two solutions I can think of:
1. This might be a bit hacky but easier: we don't change a thing and after having the expression typer type the function call we go and extract the typed patterns from the typed anonymous function knowing they're the first thing in its body
2. We change the type inference algorithm entirely to work directly on a `use` without first transforming it into a function call, this means we can type the assignments separately and not "loose" them

I hope this makes it a bit clearer, let me know if it's confusing!

Reply:
Thank you for the explanation! 1 sounds fine to me. We could add a comment explaining why we do this for future reference.

Reply:
Ok! I'll work on that then